### ISSUE TITLE
RG tag gets lost in cram format

### ISSUE THREAD

**Issue by ASLeonard** [samtools/htslib#1479]
I was initially thinking this was an issue with `samtools merge -r` as I could successfully merge several bam files into one bam file, with the respective `RG:Z:<name of cell>` appearing fine in the merged bam file, but they wouldn't appear if the final output was cram. Likewise converting with `samtools view` also lost the RG tags (and never were restored converting back, so truly lost).

I believe I narrowed down the cause to [here](https://github.com/samtools/htslib/blob/e97816474e5facdb0ef23df8be7e446eea474f24/cram/cram_encode.c#L2267) from the cram encoding
```
       // RG:Z
        if (aux[0] == 'R' && aux[1] == 'G' && aux[2] == 'Z') {
            rg = &aux[3];
            while (*aux++);
            if (CRAM_MAJOR_VERS(fd->version) >= 4)
                BLOCK_APPEND(td_b, "RG*", 3);
            continue;
        }
```
But even trying version 3.1 and version 4.0 didn't record the RG tag. I commented these lines out, and then it seemingly worked fine with my final cram file correctly having the RG tags. So unless I am missing something, this appears to be a safe/legitimate option (if not highly redundant) until RG tags are correctly stored in cram.

- Alex

**Comment by jkbonfield** [comment]
Are the RG tags in your SAM headers?

It's possible this is a corner case (I haven't explored it yet) that we didn't think of.  In CRAM RG tags are indices to the `@RG` headers, in the same way that BAM uses indices to the `@SQ` for chromosomes.

**Comment by ASLeonard** [comment]
They are no `@RG` in the headers, they are only coming from the `-r` flag in merge which
> The tag value is inferred from file names.

The RG value appears to be appearing correctly if I print it around within the aforementioned codeblock, so I guess since it doesn't exist in the index (dynamically assigned?) it doesn't get written.

It also appears the `@RG` is not written in the header by `samtools merge -r` when the output is bam, so maybe this is a samtools merge issue after all

**Comment by jkbonfield** [comment]
I can confirm a noddy SAM file with an RG line and no corresponding `@RG` header will silently lose the RG tag.

It's trivial to make it warn, but then it warns *a lot* in the case of missing headers (once per record).  Keeping track of which warnings we've emitted is complex given it's multi-threaded, so many just go with spammage as a way to repeatedly persuade people to fix their headers! :-)

I have a simple fix which can also store them verbatim (with additional warning spammage too).  I'm still unsure if this is wise or not, but it does work mostly.

**Comment by daviesrob** [comment]
It looks like the `merge` problem could be down to an incorrect way of checking for an empty `klist` [here](https://github.com/samtools/samtools/blob/0b624b9a5b/bam_sort.c#L633).  It shouldn't be too difficult to fix.

@ASLeonard can you confirm that your input files didn't have any `RG` headers or tags, so `merge` would be adding them rather than replacing?

**Comment by jkbonfield** [comment]
How were the RG tags originally created on those files?  It rather feels like "garbage in garbage out", although that's perhaps a little harsh! (Sorry)

It's hard though for merge to be populating RG headers on merged output when they don't exist in the incoming headers either.  I don't know if we even have a mechanism to scan through an entire file finding RG tags and populating the header.

So really fixing whatever produced the initial tags without adding appropriate header entries is the best fix.

**Comment by ASLeonard** [comment]
It was generated with the `-r` flag of samtools.

There are no RG tags in the individual files, but they appear in the merged -o bam file. They don't appear in the merged -o cram file. So samtools itself is adding RG tags to alignments but not adding a header line

**Comment by daviesrob** [comment]
Yes, `merge -r` definitely has a problem, no matter what other tools are doing.

**Comment by jkbonfield** [comment]
Oh my apologies, I forgot what -r did; it's creating the tags itself, so yes definitely it is the thing at fault here.  Looks like @daviesrob is already on the case.  (Thanks)

**Comment by daviesrob** [comment]
samtools/samtools#1683 is my fix for the `samtools merge` bit of this.
